feat(dropdowns): add dropdowns - INNO-637

kalinchernev opened the pull request:
Reusing expandables and link blocks, very little styles in the component itself.

yhuard commented on framework/components/ecl-dropdowns/_dropdowns.scss:
```diff
@@ -0,0 +1,22 @@
+ * @define dropdowns
+ */
+
+.ecl-dropdowns {
```
@emeryro would probably tell you "singular please" 😉

yhuard commented on framework/components/ecl-dropdowns/_dropdowns.scss:
```diff
@@ -0,0 +1,22 @@
+  position: relative;
+
+  // collapsed/open part of the dropdown
+  &__body {
```
Personally, I don't like writing thinks like `&__something` because it makes the code hard to search through. I invite you to read https://csswizardry.com/2017/02/code-smells-in-css-revisited/#string-concatenation-for-classes and to watch the screencast :)

yhuard commented on framework/components/ecl-dropdowns/_dropdowns.scss:
```diff
@@ -0,0 +1,22 @@
+  &__body {
+    background: #fff;
+    border: 1px solid #ccc;
+    border: 1px solid rgba(0, 0, 0, 0.15);
```
Why is the `border` rule defined twice?

kalinchernev commented:
@yhuard I added the links - this was easy.
For the click outside, I added a small listener in the dropdown component. I took the small polyfill from the same author, though I didn't see any big reason to add the event abstraction. Maybe I can't see a scenario for specific event handlers?

yhuard commented:
@kalinchernev the reason is the following: in the current scenario, every time we call `ECL.dropdown` we attach a new click listener on `document`. In the lib I shared, there's only one global click listener looping through a map of callbacks. It's probably safer.

yhuard commented:
I saw https://github.com/leofavre/canivete the other day, maybe we should have such helpers that we can reuse in our components

yhuard commented on framework/components/ecl-link-blocks/ecl-link-blocks.twig:
```diff
@@ -1,8 +1,51 @@
+{% endif %}
+
+{% if extraClassLinks is defined %}
+  {% set _cssClassLiks = _cssClassLiks ~ ' ' ~ extraClassLinks %}
```
`_cssClassLiks` small typo

yhuard commented on framework/components/ecl-dropdowns/ecl-dropdowns.twig:
```diff
@@ -0,0 +1,23 @@
+  } %}
+  {% include '@ec-europa/ecl-link-blocks' with {
+    'skipWrapper': true,
+    'extraClassLinks': 'ecl-dropdown__body ecl-dropdown__link-block',
```
Not very explicit :/

yhuard commented on framework/components/ecl-link-blocks/ecl-link-blocks.scss:
```diff
@@ -0,0 +1,40 @@
+  list-style: none;
+  padding-left: 0;
+
+  li {
```
Why don't you use a class here?

yhuard commented on framework/components/ecl-link-blocks/ecl-link-blocks.scss:
```diff
@@ -0,0 +1,40 @@
+    white-space: nowrap;
+  }
+
+  a:link,
```
And here?

yhuard commented on test/functional/datepickers.js:
```diff
@@ -41,7 +41,7 @@ describe('datepickers', () => {
     before(() => {
       // Reload
-      browser.url(`ecl-datepickers.html`);
+      browser.url('ecl-datepickers.html');
```
That's not necessary ^^

yhuard commented on framework/components/ecl-link-blocks/ecl-link-blocks.twig:
```diff
@@ -1,8 +1,51 @@
+        {{ extraAttributesLinks }}
+      >
+        {% for link in links %}
+          <li><a href="{{ link.target }}">{{ link.title }}</a></li>
```
Why don't the links use the `ecl-link` class?

emeryro commented:
Modifications have been done, and screenshots updated.
Performance issue will be handled in https://webgate.ec.europa.eu/CITnet/jira/browse/INNO-651
